llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -3597,6 +3607,32 @@ tryToMatchAndCreateMulAccumulateReduction(VPReductionRecipe *Red,
         dyn_cast_if_present<VPWidenCastRecipe>(B->getDefiningRecipe());
     auto *Mul = cast<VPWidenRecipe>(VecOp->getDefiningRecipe());
 
+    // Match reduce.add(mul(ext, const)) and convert it to
+    // reduce.add(mul(ext, ext(const)))
+    if (RecipeA && !RecipeB && B->isLiveIn()) {
```

Comment:
It would be nice if this could also work for the case handled on loop 3657 (`zext(mul(zext(a), zext(b)))` where `b` is a constant)